Feature/dark theme

bdmendes opened the pull request:
Closes #244 
Introduces dark theme. Switching themes is done on the hamburger menu, next to the logout button.
(the color palette needs some work) 

# Review checklist
- [ ] Terms and conditions reflect the current change
- [ ] Contains enough appropriate tests
- [x] Increments version in `pubspec.yaml` (changes `1.0.0+1` to `1.0.0+2` for example)
- [x] Properly adds entry in `changelog.md` with the change
- [x] If PR includes UI updates/additions, its description has screenshots
- [ ] Behavior is as expected
- [ ] Clean, well structured code

bdmendes commented:
Did a quick rebase just now.
The color palette is being reviewed by @ines-08. Meanwhile, I'm studying the meanings of the properties of the material design theme to make our theme data consistent and try to eliminate the use of hard coded colors across the app, which is still an issue on the login page.

bdmendes commented:
All app widgets are now visible and working. Can you please post 2 or 3 screenshots of your palettes, @ines-08 and @thePeras, so that we can vote and include it here?

ines-08 commented:
This is the latest version of what I did. If you want me to change something, please let me kno
![Screenshot_20220113_181724_pt up fe ni uni](https://user-images.githubusercontent.com/78428139/149424268-9784f3cc-3f70-4209-92f3-feb19f5ab295.jpg)
w
![Screenshot_20220113_181729_pt up fe ni uni](https://user-images.githubusercontent.com/78428139/149424281-6ef7ab3d-4ad9-460e-b072-65f293d0e883.jpg)

thePeras commented:
My idea was to keep the primary color (red) that exists in default mode. I think that makes a comfortable view and a characteristic look.

Another point I would suggest it's to make the dark even dark. That's means reducing the white overlay
transparency. What percentage number do you like most?

<img width="200" alt="dps" src="https://user-images.githubusercontent.com/53405284/149627926-e6cf8438-33ae-4e52-bd35-f42778d7de50.png">

In order to approximate the @ines-08 version, I upgraded my white transparency level. (version2)

I think that @ines-08 added a little of red dependency in the dark colors. That's not a bad idea, actually.

<div style="display: flex">
<figure style="display: flex">
 <img width="219" alt="version 1" src="https://user-images.githubusercontent.com/53405284/149627921-12bdca9a-5532-46af-ae9d-6a3f2ca28823.png">
</figure>
<figure>
<img width="223" alt="version 2" src="https://user-images.githubusercontent.com/53405284/149627924-f6216a02-cfab-495f-ba47-d8204da66583.png">
</figure>
</div>

ines-08 commented:
Yes, I added some red dependency to have some UNI colors since I didn't use red directly. 
Of two images you sent, I like the second version better. But I think the red letters give a very strong contrast.

bdmendes commented:
I'm not of the dark red on dark grey contrast. Personally, if the background is very dark, the red makes some contrast and looks nice, but on lighter backgrounds the light font might be the way to go.

andrefmrocha commented:
My biggest concern is the fact that dark grey or black and red are not color-blind friendly: https://enchroma.com/pages/types-of-color-blindness

thePeras commented:
Yes, I done some tests and black and red aren't color-blind friendly. Let's pass this idea on. The primary color isn't typically used in text but yes in box backgrounds and graphs.

I changed text to white. 
Still with a little more dark version

From @ines-08 version to my 3 black and white versions:

<div style="display: grid">
<img width="190"  src="https://user-images.githubusercontent.com/78428139/149424268-9784f3cc-3f70-4209-92f3-feb19f5ab295.jpg" alt='missing' />
<img width="200" alt="1" src="https://user-images.githubusercontent.com/53405284/149673663-be8858dd-f0f9-4c6c-a3f9-15d300ba3653.png">
<img width="200" alt="2" src="https://user-images.githubusercontent.com/53405284/149673665-ffd332ec-6109-4488-bb63-1f91aa756009.png">
<img width="200" alt="3" src="https://user-images.githubusercontent.com/53405284/149673666-50f3359f-a6a5-4c97-a4aa-706d0ee63760.png">
</div>

ines-08 commented:
Of the ones you sent now, I like mine (obviously...) and the second of yours. 
The third ones looks very dark to me, but I'm not used to using dark mode either...

tiago-falves commented:
> Yes, I done some tests and black and red aren't color-blind friendly. Let's pass this idea on. The primary color isn't typically used in text but yes in box backgrounds and graphs.
> 
> I changed text to white. Still with a little more dark version
> 
> From @ines-08 version to my 3 black and white versions:
> 
> <img alt="missing" width="190" src="https://user-images.githubusercontent.com/78428139/149424268-9784f3cc-3f70-4209-92f3-feb19f5ab295.jpg"> <img alt="1" width="200" src="https://user-images.githubusercontent.com/53405284/149673663-be8858dd-f0f9-4c6c-a3f9-15d300ba3653.png"> <img alt="2" width="200" src="https://user-images.githubusercontent.com/53405284/149673665-ffd332ec-6109-4488-bb63-1f91aa756009.png"> <img alt="3" width="200" src="https://user-images.githubusercontent.com/53405284/149673666-50f3359f-a6a5-4c97-a4aa-706d0ee63760.png">

The third image with "MD" class won with 20 votes!

thePeras commented:
The colors are:
grey: rgb(43, 43, 43)
black: rgb(27, 27, 27)
white: rgb(232, 232, 232)
dark red, the same as white mode: rgb(0x75, 0x17, 0x1e)

thePeras commented:
About the design I see no problems beside two minor things.
The 'Bugs e sugestões' checkbox doesn't notice very well in dark mode:
<img width="308" alt="Captura de ecrã 2022-03-22, às 19 04 40" src="https://user-images.githubusercontent.com/53405284/159557587-d15cd828-68b0-4ce5-bf36-04644af7fa34.png">

Another thing is the icon to add event to calendar, original was grey. To be honest red doesn't look bad but in my opinion it highlights a not primary button.
<img width="256" alt="Captura de ecrã 2022-03-22, às 19 07 27" src="https://user-images.githubusercontent.com/53405284/159557979-ed0ccb86-ed2e-48cf-ae08-3795d719e62f.png">

Another point to consider in the dark theme is to let the icons be red or white(as the PR is right now).
@ines-08 what you think?

bdmendes commented:
@thePeras I changed the provider implementation a bit and don't notice any exception when changing the theme, except a rare `Null check operator used on null value` when spamming the button. Based on [this](https://stackoverflow.com/questions/64278595/null-check-operator-used-on-a-null-value/67990442#67990442) I'm assuming this is related to some shady Flutter nullable behavior and not an error on our part.
Fixed the checkbox color.
About the icon, I made all app icons the same color to simplify and make them consistent, however I can change that if @ines-08 agrees.

thePeras commented:
> @thePeras I changed the provider implementation a bit and don't notice any exception when changing the theme, except a rare `Null check operator used on null value` when spamming the button. Based on [this](https://stackoverflow.com/questions/64278595/null-check-operator-used-on-a-null-value/67990442#67990442) I'm assuming this is related to some shady Flutter nullable behavior and not an error on our part. Fixed the checkbox color. About the icon, I made all app icons the same color to simplify and make them consistent, however I can change that if @ines-08 agrees.

The red alerts only appears on debug mode? Maybe in production is not a big issue, I think...

The checkbox is really good now.

I think the icons color can be red. There is not a big difference from the previous grey.

Just a new thing that I notice: (faculty dropdown)
<div>
<img width="200" src="https://user-images.githubusercontent.com/53405284/163169717-7a9a0685-f3d7-42c6-a9b8-2451c4fb232f.jpg" />
<img width="200" src="https://user-images.githubusercontent.com/53405284/163169722-0df89830-7ecb-46ed-b3c0-cf6189c795d4.jpg" />
</div>

In the dark mode, the background is to white.
And in the white mode, the red background is to red,

ines-08 commented:
Could you post a picture of how the icons are currently in dark mode, please?? It's just that you changed files that I also changed and they are giving conflicts...

bdmendes commented:
@ines-08 
Currently, all icons are red in the white theme, and white in the dark theme. In the white theme:
![uni_exam_icon](https://user-images.githubusercontent.com/61701401/163377486-2f90908b-a356-4505-9cdd-763acf0853d0.png)
I would prefer that all app icons have the same color, so in case you prefer the old grayish, I'd like to change them all to that. Please let me know!

ines-08 commented:
I like the red icons in white mode, but I don't know if they will go unnoticed in dark mode... 
But if you don't mind making that change and then putting an image here so we can compare, it would give us a better idea of the contrast

bdmendes commented:
@ines-08 Currently, the dark mode icons are white:
![dark_theme_white_icon](https://user-images.githubusercontent.com/61701401/163553293-e31263ea-633a-47f6-91e7-2e196f180cbd.png)
A possible alternative for the dark mode:
![dark_theme_red](https://user-images.githubusercontent.com/61701401/163553393-eb7b0b26-501a-4b92-a1f4-f34837713ebb.png)
The light theme + grey icon version you asked for:
![light_theme_grey_icon](https://user-images.githubusercontent.com/61701401/163553466-aa1ec733-aca1-493e-902b-e2e7aafa1c42.png)

ines-08 commented:
Thank you!
Now that I see them both, the red icons really look good, I like it!
For me you can leave them both in white mode and in dark mode

bdmendes commented:
@ines-08 There is little to no red on the dark theme (not even the logo), would it be ok to give the icons such highlight?

ines-08 commented:
What was your initial suggestion?? I thought it was to have the icons in red

bdmendes commented:
Red on light mode, white on dark mode.

ines-08 commented:
OK! If in dark mode not even the logo is red, then we can follow your suggestion which also looks good!
But I got confused, because I thought the goal was to have the icons always with the same color

thePeras reviewed (dismissed):
I think it's good to go!
